Feature/add structured logging interface

johannes-graeter opened the pull request:
# Description
Implemented the interface discussed in #211 and #201.
useable as 
```cpp
  const int num = 1234;
  heph::telemetry::log(heph::telemetry::Level::WARN, "testing absl log sink");
  heph::telemetry::log(heph::telemetry::Level::WARN, "testing absl log sink with field", "num", num, "quoted_string", "test");
```
see tests and examples.

## Type of change
- New feature (non-breaking change which adds functionality)

## Checklist before requesting a review
- [x] I have performed a self-review of my code.
- [x] If it is a core feature, I have added thorough tests.
- [x] If this is a new component I have added examples.
- [ ] I updated the README and related documentation.

johannes-graeter commented:
@lhruby @filippobrizzi @floriantschopp one remaining issue is if we should elevate `heph::telemetry::log` and Level to `heph::log`?

johannes-graeter commented:
For uneven parameters I added 'static_assert(false,"")' to enhance readability of the compile error in case the parameter are mixed up by the user. However the build_gcc staged failed then, but in the example in the comment above it worked fine with gcc 14.2 and cpp20.
Making this work without dirty hacks is newer according to https://stackoverflow.com/questions/76673888/why-is-always-false-v-required-in-this-situation I am AFK but I guess the gcc version in ci could be updated?

filippobrizzi commented on modules/telemetry/telemetry/include/hephaestus/telemetry/struclog.h:
```diff
@@ -0,0 +1,187 @@
+  }
+
+  Level level;
+  MessageWithLocation msg_with_loc;
```
Can we unpack message and location here?

filippobrizzi commented on modules/telemetry/telemetry/include/hephaestus/telemetry/struclog.h:
```diff
@@ -0,0 +1,187 @@
+}
+
+/// @brief Interface for sinks that log entries can be sent to.
+struct ILogSink {
+  using Formatter = std::function<std::string(const LogEntry& log)>;
+
+  virtual ~ILogSink() = default;
+
+  /// @brief Main interface to the sink, gets called on log.
+  ///
+  /// @param log_entry
+  virtual void send(const LogEntry& log_entry) = 0;
+};
```
please move this into `log_sink.h`

johannes-graeter replied:
done for this and LogEntry (same as metric_sink and metric_record) also renamed struclog.h to log.h

filippobrizzi commented on modules/telemetry/telemetry/src/log_sinks/absl_sink.cpp:
```diff
@@ -0,0 +1,44 @@
+      break;
+  }
+}
+}  // namespace heph::telemetry
```
missing end-of-line

filippobrizzi commented on modules/telemetry/telemetry/include/hephaestus/telemetry/struclog.h:
````diff
@@ -0,0 +1,187 @@
+///       ```
+///       log(Level::WARN, "speed is over limit", "current_speed", 31.3, "limit", 30.0, "entity", "km/h")
+///       ```
+template <typename... Args>
+void log(Level level, MessageWithLocation msg, Args&&... fields) {
+  internal::logWithFields(LogEntry{ level, msg }, std::forward<Args>(fields)...);
+}
+
+///@brief Log a message without fields. Example:
+///       ```
+///       log(Level::WARN, "speed is over limit", "current_speed"))
+///       ```
+template <typename... Args>
+void log(Level level, MessageWithLocation msg) {
+  internal::log(LogEntry{ level, msg });
+}
````
let's place this in `namespace heph`

filippobrizzi commented:
> For uneven parameters I added 'static_assert(false,"")' to enhance readability of the compile error in case the parameter are mixed up by the user. However the build_gcc staged failed then, but in the example in the comment above it worked fine with gcc 14.2 and cpp20. Making this work without dirty hacks is newer according to https://stackoverflow.com/questions/76673888/why-is-always-false-v-required-in-this-situation I am AFK but I guess the gcc version in ci could be updated?

We could add 
```c++
#ifndef __GNUC__
// TODO: remove ifdef once we migrate to GCC 14
...
#endf
```

filippobrizzi reviewed (commented):
Fantastic work!! please check my comments

johannes-graeter commented:
I also elevated the Log level as to write less to `heph::Error`is that too ambiguitive? perhaps we can have a short namespace in between like heph::loglevel::Error or use a name like `heph::ErrorLevel`or `heph::ErrorL`
@filippobrizzi WDYT?

filippobrizzi commented:
> I also elevated the Log level as to write less to `heph::Error`is that too ambiguitive? perhaps we can have a short namespace in between like heph::loglevel::Error or use a name like `heph::ErrorLevel`or `heph::ErrorL` @filippobrizzi WDYT?

What about `heph::LogLevel::INFO`?

johannes-graeter commented:
> > For uneven parameters I added 'static_assert(false,"")' to enhance readability of the compile error in case the parameter are mixed up by the user. However the build_gcc staged failed then, but in the example in the comment above it worked fine with gcc 14.2 and cpp20. Making this work without dirty hacks is newer according to https://stackoverflow.com/questions/76673888/why-is-always-false-v-required-in-this-situation I am AFK but I guess the gcc version in ci could be updated?
> 
> We could add
> 
> ```c++
> #ifndef __GNUC__
> // TODO: remove ifdef once we migrate to GCC 14
> ...
> #endf
> ```

I have added `#if (__GNUC__ >= 14) || defined(__clang__)`

johannes-graeter commented:
One thing that makes me still a little uneasy is that now the message is not owned by LogEntry anymore. For literals this doesn't not matter as stated above since they are static, but it would be easy to get UB if you would decide to compose the message before as string and give over the c_str... this is not the usual usage, but could still happen.

It would be safer to store the message as std::string, but I am uncertain if the compiler would inline the static string literal in normal usage or if it would copy...

Do you know that @filippobrizzi or @lievenvanderheide ?

johannes-graeter commented:
> > One thing that makes me still a little uneasy is that now the message is not owned by LogEntry anymore. For literals this doesn't not matter as stated above since they are static, but it would be easy to get UB if you would decide to compose the message before as string and give over the c_str... this is not the usual usage, but could still happen.
> 
> > 
> 
> > It would be safer to store the message as std::string, but I am uncertain if the compiler would inline the static string literal in normal usage or if it would copy. For large strings I am pretty certain that it will copy but for short ones, I don't know...
> 
> > 
> 
> > Do you know that @filippobrizzi or @lievenvanderheide ?
> 
> 
> 
> The reason why `MessageWithLocation` doesn't save a `std::string` is to avoid copies? 

So the constructor of MessageWithLocation needs a const char* as input otherwise "my message" will not implicitly casted to it. 
It does not store as string to avoid copies yes.
I guess if we use "my message"s we could have a string in the constructor... we could the message type everywhere and enforce << through a concept but I am unsure if that is not overengeneered...

floriantschopp commented on modules/telemetry/telemetry/examples/log_sink_example.cpp:
```diff
@@ -0,0 +1,23 @@
+  heph::log(heph::LogLevel::WARN, "testing absl log sink");
+
+  const int num = 1234;
+  heph::log(heph::LogLevel::INFO, "absl log sink with fields", "num", num, "quoted_string", "test");
```
great, thanks guys for adjusting, how about the fields being pairs, ie.e 
```suggestion
  heph::log(heph::LogLevel::INFO, "absl log sink with fields", {"num", num}, {"quoted_string", "test"});
```
makes it clearer to what belongs together in a key-value way, no?

johannes-graeter replied:
yes definitely, this was our prefered solution, however I struggled combining brace initializers with variadics...
see https://github.com/olympus-robotics/hephaestus/pull/211
what your take, would you prefer
```
heph::log(heph::LogLevel::INFO, "absl log sink with fields", {{"num", num}, {"quoted_string", "test"}});
```
? @filippobrizzi and I prefered the interface with less to write...

floriantschopp replied:
thats ok, lets keep the above

johannes-graeter commented:
@filippobrizzi I added an additional constructor supporting std::string to MessageWithLocation.
This has 2 advantages:
A. using string literals ""s we can use the interface without copying in case we have long strings
B. we can use log(LogLevel::Warn, std::format("{}",123)); without copies (see https://github.com/olympus-robotics/hephaestus/pull/201#discussion_r1834707786)

johannes-graeter commented:
since commit [2bad994](https://github.com/olympus-robotics/hephaestus/pull/213/commits/2bad994d308c248d98bb0a206373c7550bb2b786) I am struggling with the fact that fmt/core.h seems deprecated now. Looks as if https://github.com/olympus-robotics/hephaestus/pull/214 solved this. 
If some fix gets merged before this, we should reset to [8479ea0](https://github.com/olympus-robotics/hephaestus/pull/213/commits/8479ea0ed2009b7136e4347c2ee856caa6ea6ed8) and then merge main.

johannes-graeter commented:
blocked by https://github.com/olympus-robotics/hephaestus/pull/214

filippobrizzi commented on modules/telemetry/telemetry/include/hephaestus/telemetry/log_sink.h:
```diff
@@ -0,0 +1,145 @@
+#include <vector>
+
+namespace heph {
+enum class LogLevel : std::uint8_t { TRACE, DEBUG, INFO, WARN, ERROR, FATAL };
```
Am I crazy if I suggest:
```suggestion
enum LogLevel : std::uint8_t { TRACE, DEBUG, INFO, WARN, ERROR, FATAL };
```
so that we can do: `heph::log(heph::INFO, ...)` 

This symbols are very log specific, I doubt we will cause any clashes

johannes-graeter replied:
I actually also considered it, however I am realy not a fan of implicit cast of enums to uchar...
we could consider strong typing, just defining them as individual structs and enforcing them with a concept like
https://godbolt.org/z/nG1G4PrGb
might be overkill though

filippobrizzi replied:
This is a cool concept, but I think we can live with having a non class enum. It is for a very specific case, so it should be ok.

filippobrizzi commented on modules/telemetry/telemetry/src/log_sink.cpp:
```diff
@@ -0,0 +1,77 @@
+}
+
+auto format(const LogEntry& log) -> std::string {
+  std::stringstream ss;
+  ss << "level=" << log.level;
+  ss << " hostname=" << std::quoted(log.hostname);
+  ss << " location="
+     << std::quoted(fmt::format("{}:{}",
+                                std::filesystem::path{ log.location.file_name() }.filename().string(),
+                                log.location.line()));
+  ss << " thread-id=" << log.thread_id;
+  ss << " time=" << fmt::format("{:%Y-%m-%dT%H:%M:%SZ}", log.time);
+  ss << " message=" << std::quoted(log.message);
+
+  for (const Field<std::string>& field : log.fields) {
+    ss << " " << field.key << "=" << field.value;
+  }
+
+  return ss.str();
```
nit. I am wondering if it is not better to use `fmt::format`, I know you cannot use `std::quoted` but it should be still easy to add them with `\"\"` and it is more inline with the rest of the formatting in hephaestus.

Also for field you can use `join`.

johannes-graeter replied:
adding  ` \"\"` is unfortuanetly not the same behaviour, since std::quoted would not quote the string if it is already quoted (to get rid of double quotes). So we would need to check if the user quoted the quotable which is also easy enough but would add some boiler plate. Also we can just life with the fact that the user may not quote the message or gets double quotes...

johannes-graeter replied:
i now added fmt for everything, also replaced the format() function by a fmt::formatter to be consistent (using fmt as default for formatting everywhere...).
only downside is that the formatting now all resides in the header... what do you think? @filippobrizzi

johannes-graeter commented on modules/telemetry/telemetry/include/hephaestus/telemetry/log.h:
```diff
@@ -0,0 +1,75 @@
+
+namespace heph::telemetry {
+
+void registerLogSink(std::unique_ptr<ILogSink> sink);
```
so in the heph namespace right?

filippobrizzi commented on modules/telemetry/telemetry/include/hephaestus/telemetry/log.h:
```diff
@@ -0,0 +1,73 @@
+}
+
+///@brief Register a sink for logging.
+void registerLogSink(std::unique_ptr<telemetry::ILogSink> sink);
```
sorry if it wasn't clean I would keep this here, but maybe I would still wrap it in `heph::telemetry` namespace.

filippobrizzi commented on modules/telemetry/telemetry/include/hephaestus/telemetry/log_sink.h:
```diff
@@ -0,0 +1,171 @@
+};
+
+template <>
+struct fmt::formatter<heph::LogLevel> : fmt::formatter<std::string_view> {
+  static auto format(const heph::LogLevel& level, fmt::format_context& ctx) {
+    switch (level) {
+      case heph::LogLevel::TRACE:
+        return fmt::format_to(ctx.out(), "trace");
+      case heph::LogLevel::DEBUG:
+        return fmt::format_to(ctx.out(), "debug");
+      case heph::LogLevel::INFO:
+        return fmt::format_to(ctx.out(), "info");
+      case heph::LogLevel::WARN:
+        return fmt::format_to(ctx.out(), "warn");
+      case heph::LogLevel::ERROR:
+        return fmt::format_to(ctx.out(), "error");
+      case heph::LogLevel::FATAL:
+        return fmt::format_to(ctx.out(), "fatal");
+      default:
+        return fmt::format_to(ctx.out(), "XXX");
+    }
+  }
```
why not using `magic_enum`?

johannes-graeter replied:
oh yeah awesome!

filippobrizzi reviewed (commented):
I think we are ready! last 2 comments and then we can merge!

johannes-graeter commented:
Added suggestions and ported to bazel. Took the liberty to add the metric_sink_example as binary.

filippobrizzi commented on modules/telemetry/telemetry/BUILD:
```diff
@@ -35,3 +36,31 @@ heph_cc_test(
+    ],
+)
+
+############
```
Very nit. but I wouild put binaries before tests, so that it's easier to see them

filippobrizzi reviewed (approved):
Fantastic work!

johannes-graeter commented:
started adding the logs in branch feature/port-to-hephlog
